### PR TITLE
Rewrite typing tests as assert_type statements

### PR DESCRIPTION
**Pull request by ikonst**
With the advent of [assert_type](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-types-in-assert-type-assert-type), we can adopt a [testing approach](https://typing.readthedocs.io/en/latest/source/quality.html#testing-using-mypy-warn-unused-ignores) that's a little less verbose but no longer needing any "special sauce".

### CODE REVIEW

**Comment by ikonst** [comment]
Have to merge in #1057

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -31,6 +31,11 @@ jobs:
         tar -xzf /tmp/dynamodb_local_latest.tar.gz -C /tmp
         java -Djava.library.path=/tmp/DynamoDBLocal_lib -jar /tmp/DynamoDBLocal.jar -inMemory -port 8000 &
 
+    - name: Run tests
```
Comment (garrettheel) [inline .github/workflows/test.yaml:34]: Is this an accidental duplicate?

**Review by garrettheel — APPROVED** [review]
nice!

**Review by NazarioJL — COMMENTED** [review]
This is awesome! 👍

**Pull request merged**